### PR TITLE
[FIX] hw_drivers: inform user on printing failed

### PR DESCRIPTION
**Pull request by Jihaysse**
Currently, the user does not get any feedback if the printer couldn't print.
This can happen when the printer identifier is not the same on CUPS and on the IoT box due to this RegEx:
https://github.com/odoo/odoo/blob/5b62e301b91a3f39857c2c7f18729e10796ef194/addons/hw_drivers/iot_handlers/interfaces/PrinterInterface.py#L55

After this commit, a pop-up error is displayed in the PoS as well as an error message in the IoT box logs so the user is informed that the printing failed.
Also, the '`-`' character has been added to the RegEx as it is a commonly used character in printer's identifier as CUPS doesn't allow spaces and nothing seems to justify filtering it.

opw-3458347

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Comment by lse-odoo** [comment]
Note for QLE: this a newbie in our team. I will do the initial review as it is his first PRs

File: addons/hw_drivers/controllers/driver.py [hunk 1]
```diff
@@ -40,8 +40,7 @@ def action(self, session_id, device_identifier, data):
                     _logger.info("Ignored request from %s as iot_idempotent_id %s already received from session %s",
                                  session_id, iot_idempotent_id, idempotent_session)
                     return False
-            iot_device.action(data)
-            return True
+                return iot_device.action(data) if not None else True
```
Comment (lse-odoo) [inline addons/hw_drivers/controllers/driver.py:43]: what is the purpose of this part ?

I guess that you try to don't always return `True` if the function return something. But I am a bit confused.

I am also a bit afraid that it will have side effects:
let's assume that `iot_device.action(data)` return `None`, in this case, with your code it will also return `None` which I doubt it is intended. In general we prefer to use `False` if something fail.
Note that in python you can use `or` and `and` to easily alter the value of boolean

File: addons/hw_drivers/iot_handlers/drivers/PrinterDriver.py [hunk 2]
```diff
@@ -165,6 +165,12 @@ def send_status(self):
     def print_raw(self, data):
         process = subprocess.Popen(["lp", "-d", self.device_identifier], stdin=subprocess.PIPE)
         process.communicate(data)
+        
+        if process.returncode == 0:
+            return True
+        # The stderr isn't meaningful so we don't log it ('No such file or directory')
+        _logger.error('Error while printing: please check if "%s" matches the name provided on CUPS.', self.device_identifier)
```
Comment (lse-odoo) [inline addons/hw_drivers/iot_handlers/drivers/PrinterDriver.py:172]: The intention is nice, but I don't think it would be a good idea to have this kind of error message.
Let's assume we don't use CUPS anymore (or that we use it for Windows IoT box, which does not use CUPS), your message will be misleading.
I think it would be better to just point the fact that we know. Something like:
`Printing failed: printer with the identifier "%s" could not be found`

File: addons/hw_drivers/iot_handlers/interfaces/PrinterInterface.py [hunk 3]
```diff
@@ -52,9 +52,9 @@ def get_devices(self):
 
     def get_identifier(self, path):
         if 'uuid=' in path:
-            identifier = sub('[^a-zA-Z0-9_]', '', path.split('uuid=')[1])
+            identifier = sub('[^a-zA-Z0-9_-]', '', path.split('uuid=')[1])
```
Comment (lse-odoo) [inline addons/hw_drivers/iot_handlers/interfaces/PrinterInterface.py:55]: may be a good occasion to extract this string into a constant prior to the `if` rather than this copy paste ?

File: addons/point_of_sale/static/src/js/printers.js [hunk 4]
```diff
@@ -30,7 +30,7 @@ class PrintResultGenerator {
             successful: false,
             message: {
                 title: _t('Connection to the printer failed'),
-                body: _t('Please check if the printer is still connected. \n' +
+                body: _t('Please check if the printer is still connected and make sure there are no errors in the IoT box logs. \n' +
```
Comment (lse-odoo) [inline addons/point_of_sale/static/src/js/printers.js:33]: Once again, the intention is nice but it can be misleading as well.
In the sense that if we assumed that they have a ton of unrelated errors in their log files (for example because the initial setup was done badly), they might be confused and try to solve every errors in their logs which are not related to the issue.
Maybe we should replace it with something like: `your IoT may have logged more information regarding printing errors` (intentionally use words that you set in the logger error).

Comment (lse-odoo) [inline addons/point_of_sale/static/src/js/printers.js:33]: Something else that you have to be careful of is the fact that this string is in a `_t(...)` (which is the python equivalent of `_()`.
Which means that this sentence is meant to be translated.

Here there is 2 issues with your change:
- You modify a string that might have existing translations (so they will be lost).
- You didn't update the .pot file with your new text

=> You will prevent the translation until someone update it for you & the previous translation will be lost

In this case you are a bit in trouble as the full line is translated as we can see in:
https://github.com/odoo/odoo/blob/5eeb565fb094ee4f51994bafceb2110760081228/addons/point_of_sale/i18n/point_of_sale.pot#L4124-L4126
But, what you could do is for instance appending at the end of the `_t(...)` your message. Like: 
```js
body: _t('Please check if the printer is still connected. \n' + 
...
'\'Self-signed certificate for ePOS printers\' and \'Secure connection (HTTPS)\' to solve the issue'
 ) 
+ "\n" 
+  _t("make sure there are no errors in the IoT box logs."),
```

Comment (lse-odoo) [inline addons/point_of_sale/static/src/js/printers.js:33]: Maybe we should rethink this part.
The error here assumes that the printer was not reachable (the request itself failed to be received by the target (Iot or printer) ), in our case, the IoT box was reached AND did replied. Can't we make a distinction as we did receive something to have a more relevant error message ?

I mean, let's assume that a customer receive this error message, it's not obvious what they should do as we give ton of instructions. I think it would really help them if you can have your own custom error message at a more relevant place in the code.

**Review by lse-odoo — CHANGES_REQUESTED** [review]

File: addons/hw_drivers/controllers/driver.py (hunk 1 above)
Comment (Jihaysse) [reply addons/hw_drivers/controllers/driver.py:43]: The intent is indeed to keep the default behaviour: return True (as before this commit) unless `iot_device.action(data)` returned a value.
From my understanding, there are 3 cases:
- `iot_device.action(data)` returns True
- `iot_device.action(data)` returns False
- `iot_device.action(data)` returns None => True will be returned instead

Excuse the newbie Python from my end, I think this will make more sense: 
```
result = iot_device.action(data)
return result if result is not None else True
```

The reason I want to handle the case where the value returned is `None` as no actions returns a value except for actions related to printing.

File: addons/hw_drivers/iot_handlers/drivers/PrinterDriver.py (hunk 2 above)
Comment (Jihaysse) [reply addons/hw_drivers/iot_handlers/drivers/PrinterDriver.py:172]: I totally agree with this, thanks for the input!

File: addons/point_of_sale/static/src/js/printers.js [hunk 5]
```diff
@@ -25,16 +25,25 @@ class PrintResultGenerator {
             },
         });
     }
-    IoTResultError() {
+    IoTResultError(printResult) {
+        var body;
+        if (printResult.identifier === undefined)
+        {
+            body = _t('Cannot find a printer.\nYour IoT box may have logged more information concerning this error.');
```
Comment (lse-odoo) [inline addons/point_of_sale/static/src/js/printers.js:32]: `Cannot find a printer` is ambiguous here, an IoT can have several printers. Can't you retrieve with the current context the name of the printer maybe ? If not, maybe prefer a formulation like:
"Cannot find the printer"
or if we know that it is necessarily the receipt printer:
"Cannot find the receipt printer"

File: addons/point_of_sale/static/src/js/printers.js [hunk 6]
```diff
@@ -77,7 +86,7 @@ var PrinterMixin = {
             // IoT box can't find a printer.
             if (!sendPrintResult || sendPrintResult.result === false) {
                 this.receipt_queue.length = 0;
-                return this.printResultGenerator.IoTResultError(sendPrintResult.printerErrorCode);
+                return this.printResultGenerator.IoTResultError(sendPrintResult);
```
Comment (lse-odoo) [inline addons/point_of_sale/static/src/js/printers.js:89]: why was this changed ? i do wonder if it will not alter other flows

**Review by lse-odoo — CHANGES_REQUESTED** [review]

File: addons/point_of_sale/static/src/js/printers.js (hunk 6 above)
Comment (Jihaysse) [reply addons/point_of_sale/static/src/js/printers.js:89]: I've changed this part because I need the whole `sendPrintResult` object to check if the `sendPrintResult.idenrtifier === undefined` (printers.js, L30). You're right, it may break the IoTResultError() override in /home/odoo/src/odoo/addons/pos_epson_printer/static/src/js/printers.js. I'm fixing this in my next commit, thanks! :1st_place_medal:

Comment (lse-odoo) [reply addons/point_of_sale/static/src/js/printers.js:87]: @Jihaysse thanks for the details, indeed you will need it and your attempt to fix it is nice, however, there is a catch here.
If we were on master you could have proposed that, but as we are in stable version we should NOT change function arguments. Think about it, if someone override this function with custom code, the argument will now suddenly change on his side. Even if it is pretty unlikely in this case, it can happens. We will likely discuss it with the PoS/IoT team later on. But I guess it would be better in this case to add your parameter as an extra argument:
```suggestion
                return this.printResultGenerator.IoTResultError(sendPrintResult.printerErrorCode, sendPrintResult);
```
(and adapt below to use the second argument)

We do agree that it is a bit stupid as the second argument contains the first one. But this is the best tradeoff I can think of if we want to ensure stability

**Comment by lse-odoo** [comment]
@Jihaysse 
I did tested for real your code on a printer. It did give the intended result. However, I did notice a few things that might be worth mentioning:

 A. Updating the code will create duplicate printers with the identifier that change
image
the PoS will show the printer as "disconnected" if we don't modify it.
If we want to change it they have the same name so it's ambiguous which one should be used

I don't think we can do much on the matter, except maybe modifying the domain or device name to show online/offline ?

 B. As you modify iot handlers code and hw_drivers code, it's important to be careful as the iot handlers code is very likely to be updated without the hw_drivers code. I didn't noticed anything terrible in term of stability but some weird behaviors that might be worth to double check.

 C. Let's assume that the iot handlers are updated with your code but not hw_drivers. If I try to print with a printer that have an unused identifier (you can just modify manually a real PoS used printer identifier manually in the Iot devices). Then your error message will pop-up (which happens as https://github.com/odoo/odoo/blob/92e84e3c78908a95aad6b76c2bc39add0e161430/addons/hw_drivers/controllers/driver.py#L30-L45 will return `False` as `iot_device` is set to `None` as the identifier is not in `iot_devices`. Is it intended to have your message in this situation ?

**Comment by lse-odoo** [comment]
( details on what have been tested: )

1. create a printer with a "-" in the name + without branch -> return true in response and nothing happen
2. branch applied -> nothing changed
3. handlers reloaded -> create duplicated device name in odoo devices and other printers set as offline and disconnected in PoS
4. manually delete the printer & modify the receipt printer for the new one -> print as intended
5. remove the printer from cups (to force using an identifier that the IoT loaded but do not exist any longer) -> error in the log but not in PoS
6. restart IoT odoo service -> error message in PoS "Cannot find the printer" (which happen NOT due to this PR but to the fact that `iot_devices.get(device_identifier)` do not find the identifier any longer)
7. force update the code of hw_drivers (as NOT iot_handlers) -> does show the error message if printer used then deleted from CUPS

**Comment by Jihaysse** [comment]
Thanks for your investigation @lse-odoo.

> A. Updating the code will create duplicate printers with the identifier that change. I don't think we can do much on the matter, except maybe modifying the domain or device name to show online/offline ?

I agree, we can either modify the domain of `pos.config` `iface_printer_id` (e.g. `('connected', '=', True)`) or change the device name. In my opinion, if I've fully understood, both cases have culprits:
- Modifying the domain to only show connected printers may be an unexpected behaviour for the customer as we show all devices of other types, connected or not. Also, they may want to select a printer which is currently disconnected and not understand why it isn't showing.
- Changing the device name (e.g. adding "[OFFLINE]" at the end of the actual name) would mean seeing that OFFLINE tag in every form/view as we need to change the name in `iot.device` and may lead to a duplicate of information, as for example in this form where there is already a checkbox indicating if the device is connected or not: 
image

Of course, we can also add the `('connected', '=', True)` domain to the other device types but it seems like a harsh modification for the end user.

Out of these 2 solutions, I would settle for changing the domain (only for the printers) as it seems to have the least impact while still solving the initial issue. What's your opinion on the matter?

> B. As you modify iot handlers code and hw_drivers code, it's important to be careful as the iot handlers code is very likely to be updated without the hw_drivers code. I didn't noticed anything terrible in term of stability but some weird behaviors that might be worth to double check.

I've checked but didn't find an unexpected issue which may arise due to these modifications.

> C. Let's assume that the iot handlers are updated with your code but not hw_drivers. If I try to print with a printer that have an unused identifier (you can just modify manually a real PoS used printer identifier manually in the Iot devices). Then your error message will pop-up (which happens as [...] will return False as iot_device is set to None as the identifier is not in iot_devices. Is it intended to have your message in this situation ?

In my opinion, in the case it wouldn't print the file, it's indeed intended to get the error pop-up.
Did it print the file in this use case?

Thank you for your time, I appreciate it!

**Comment by lse-odoo** [comment]
@Jihaysse 
A. I guess we may just ignore this and assume that client will remove the offline printers manually if they want, like you say, both ways have cons that are not great :/ 

B. Ok, we'll see if the IoT team notice stuff that we might have missed

C. I am not sure, but I think it did not

Thank you for your time, I will add the IoT team to review the PR :)

File: addons/hw_drivers/controllers/driver.py [hunk 7]
```diff
@@ -40,8 +40,8 @@ def action(self, session_id, device_identifier, data):
                     _logger.info("Ignored request from %s as iot_idempotent_id %s already received from session %s",
                                  session_id, iot_idempotent_id, idempotent_session)
                     return False
-            iot_device.action(data)
-            return True
+            result = iot_device.action(data)
+            return result if result is not None else True
```
Comment (qle-odoo) [inline addons/hw_drivers/controllers/driver.py:44]: this function `action(self, session_id, device_identifier, data)` aims to return the request action to the correct device. So what this function should return is "did you send to the correct device yes or no"
At no time should it return anything other than this response.
Here it returns the response from the driver function.

Have you looked at how others drivers return information in Odoo?

**Comment by lse-odoo** [comment]
@Jihaysse as @qle-odoo mentioned, the return-chain is something that was unintended by design, so we will have to get rid of it unfortunately :/
There is a way to solve our issue but it need to modify the code structure, replying on the IoT longpolling to reply to the request but it is trickier. Depending on your time and how much this PR interest you, let me know what you would prefer. If it does not interest you this much, just keep the logging part on the IoT box.

**Comment by lse-odoo** [comment]
@qle-odoo could you review this PR please ? :)
